transaction/src/plan/build.rs

```diff
@@ -163,4 +163,186 @@ impl TransactionPlan {
+    pub async fn build_concurrent<R: CryptoRng + RngCore>(
+        self,
+        rng: &mut R,
+        fvk: &FullViewingKey,
+        auth_data: AuthorizationData,
+        witness_data: WitnessData,
+    ) -> Result<Transaction> {
+        // Do some basic input sanity-checking.
+        let spend_count = self.spend_plans().count();
+        if auth_data.spend_auths.len() != spend_count {
+            return Err(anyhow::anyhow!(
+                "expected {} spend auths but got {}",
+                spend_count,
+                auth_data.spend_auths.len()
+            ));
+        }
+
+        let mut fmd_clues = Vec::new();
+        let mut synthetic_blinding_factor = Fr::zero();
+
+        // Add the memo.
+        let mut memo: Option<MemoCiphertext> = None;
+        let mut memo_key: Option<PayloadKey> = None;
+        if self.memo_plan.is_some() {
+            let memo_plan = self.memo_plan.clone().unwrap();
+            memo = memo_plan.memo().ok();
+            memo_key = Some(memo_plan.key);
+        }
+
+        // We build the actions sorted by type, with all spends first, then all
+        // outputs, etc.  This order has to align with the ordering in
+        // TransactionPlan::effect_hash, which computes the auth hash of the
+        // transaction we'll build here without actually building it.
+
+        // Start building the transaction's spends.
+        let mut in_progress_spend_actions = Vec::new();
+        for (spend_plan, auth_sig) in self
+            .spend_plans()
+            .cloned()
+            .zip(auth_data.spend_auths.into_iter())
+        {
+            let note_commitment = spend_plan.note.commit();
+            let auth_path = witness_data
+                .state_commitment_proofs
+                .get(&note_commitment)
+                .context(format!("could not get proof for {:?}", note_commitment))?
+                .clone();
+
+            synthetic_blinding_factor += spend_plan.value_blinding;
+            let fvk_ = fvk.clone();
+            in_progress_spend_actions.push(tokio::spawn(async move {
+                spend_plan.spend_concurrent(fvk_, auth_sig, auth_path).await
+            }));
+        }
+
+        // Start building the transaction's outputs.
+        let mut in_progress_output_actions = Vec::new();
+        let dummy_payload_key: PayloadKey = [0u8; 32].into();
+        // If the memo_key is None, then there is no memo, and we populate the memo key
+        // field with a dummy key.
+        for output_plan in self.output_plans().cloned() {
+            // Outputs subtract from the transaction's value balance.
+            synthetic_blinding_factor += output_plan.value_blinding;
+            let ovk = fvk.outgoing().clone();
+            let memo_key = memo_key.as_ref().unwrap_or(&dummy_payload_key).clone();
+            in_progress_output_actions.push(tokio::spawn(async move {
+                output_plan.output_concurrent(&ovk, &memo_key).await
+            }));
```

Comment:
in re: https://github.com/penumbra-zone/penumbra/pull/1970/files#r1102024216 , couldn't this just be
```suggestion
            in_progress_output_actions.push(tokio::spawn(async move {
                output_plan.output(&ovk, &memo_key)
            }));
```

Reply:
ah! totally, it should be